base/linalg/rowvector.jl

```diff
@@ -0,0 +1,214 @@
+
+@propagate_inbounds getindex(rowvec::RowVector, i) = transpose(rowvec.vec[i])
+@propagate_inbounds setindex!(rowvec::RowVector, v, i) = setindex!(rowvec.vec, transpose(v), i)
+
+# Cartesian indexing is distorted by getindex
+# Furthermore, Cartesian indexes don't have to match shape, apparently!
+@inline function getindex(rowvec::RowVector, i::CartesianIndex)
+    @boundscheck if !(i.I[1] == 1 && i.I[2] ∈ indices(rowvec.vec)[1] && check_tail_indices(i.I...))
+        throw(BoundsError(rowvec, i.I))
+    end
+    @inbounds return transpose(rowvec.vec[i.I[2]])
+end
+@inline function setindex!(rowvec::RowVector, v, i::CartesianIndex)
+    @boundscheck if !(i.I[1] == 1 && i.I[2] ∈ indices(rowvec.vec)[1] && check_tail_indices(i.I...))
+        throw(BoundsError(rowvec, i.I))
+    end
+    @inbounds rowvec.vec[i.I[2]] = transpose(v)
+end
+
+@propagate_inbounds getindex(rowvec::RowVector, ::CartesianIndex{0}) = getindex(rowvec)
+@propagate_inbounds getindex(rowvec::RowVector, i::CartesianIndex{1}) = getindex(rowvec, i.I[1])
+
+@propagate_inbounds setindex!(rowvec::RowVector, v, ::CartesianIndex{0}) = setindex!(rowvec, v)
+@propagate_inbounds setindex!(rowvec::RowVector, v, i::CartesianIndex{1}) = setindex!(rowvec, v, i.I[1])
+
+@inline check_tail_indices(i1, i2) = true
+@inline check_tail_indices(i1, i2, i3, is...) = i3 == 1 ? check_tail_indices(i1, i2, is...) :  false
+
+# helper function for below
+@inline to_vec(rowvec::RowVector) = transpose(rowvec)
+@inline to_vec(x::Number) = x
+@inline to_vecs(rowvecs...) = (map(to_vec, rowvecs)...)
+
+# map
+@inline map(f, rowvecs::RowVector...) = RowVector(map(f, to_vecs(rowvecs...)...))
+
+# broacast (other combinations default to higher-dimensional array)
+@inline broadcast(f, rowvecs::Union{Number,RowVector}...) = RowVector(broadcast(f, to_vecs(rowvecs...)...))
+
+# Horizontal concatenation #
+
+@inline hcat(X::RowVector...) = transpose(vcat(map(transpose, X)...))
+@inline hcat(X::Union{RowVector,Number}...) = transpose(vcat(map(transpose, X)...))
+
+@inline typed_hcat{T}(::Type{T}, X::RowVector...) = transpose(typed_vcat(T, map(transpose, X)...))
+@inline typed_hcat{T}(::Type{T}, X::Union{RowVector,Number}...) = transpose(typed_vcat(T, map(transpose, X)...))
+
+# Multiplication #
+
+@inline function *(rowvec::RowVector, vec::AbstractVector)
+    if length(rowvec) != length(vec)
+        throw(DimensionMismatch("A has dimensions $(size(rowvec)) but B has dimensions $(size(vec))"))
+    end
+    sum(@inbounds(return rowvec[i]*vec[i]) for i = 1:length(vec))
+end
+@inline *(rowvec::RowVector, mat::AbstractMatrix) = transpose(mat.' * transpose(rowvec))
+*(vec::AbstractVector, mat::AbstractMatrix) = throw(DimensionMismatch("Cannot left-multiply a matrix by a vector")) # Should become a deprecation
+*(::RowVector, ::RowVector) = throw(DimensionMismatch("Cannot multiply two transposed vectors"))
+@inline *(vec::AbstractVector, rowvec::RowVector) = kron(vec, rowvec)
```

Comment:
I think you should avoid using `kron` since it is just a matrix-matrix multiplication when the inputs are vectors. It is more likely that matrix packages have matrix-matrix multiplication implemented than `kron`.